File: bench/bigmatrix/src/mandel.cpp
/* Shape 4: mandel -- C++ port of src/mandel.c.
 *
 * Read `w h maxiter`; same loop as bench/heavy/mb.c (double, escape at |z|>2).
 * The arithmetic is deliberately written with the reference's exact types and
 * operation order: the escape test sits on a knife edge, so reassociating or
 * promoting anything here would change the iteration counts. */
#include <cstddef>
#include <cstdint>
#include <cstdio>
#include <memory>
#include <vector>

namespace {

/* Fast stdin: 1 MiB fread buffer + hand-written integer parsing, mirroring
 * src/fastio.h, so every shape reads stdin the same way. */
class StdinReader {
public:
    /* Signed parse (fastio.h bm_i64): skips ' ', '\n', '\t', '\r'. */
    std::int64_t readI64() noexcept {
        int c = get();
        bool neg = false;
        std::int64_t v = 0;
        while (c == ' ' || c == '\n' || c == '\t' || c == '\r') c = get();
        if (c == '-') {
            neg = true;
            c = get();
        }
        while (c >= '0' && c <= '9') {
            v = v * 10 + (c - '0');
            c = get();
        }
        return neg ? -v : v;
    }

    /* Unsigned parse (fastio.h bm_u64): skips anything that is not a digit and
     * yields 0 at end of input. */
    std::uint64_t readU64() noexcept {
        int c = get();
        std::uint64_t v = 0;
        while (c < '0' || c > '9') {
            if (c < 0) return 0;
            c = get();
        }
        while (c >= '0' && c <= '9') {
            v = v * 10 + static_cast<std::uint64_t>(c - '0');
            c = get();
        }
        return v;
    }

private:
    static constexpr std::size_t kBufSize = 1u << 20;

    /* Returns the next byte, or -1 at end of input. */
    int get() noexcept {
        if (pos_ == len_) {
            len_ = std::fread(buf_.get(), 1, kBufSize, stdin);
            pos_ = 0;
            if (len_ == 0) return -1;
        }
        return buf_[pos_++];
    }

    /* new[] rather than std::make_unique<unsigned char[]>: make_unique
     * value-initialises, i.e. memsets all 1 MiB that fread overwrites anyway. */
    std::unique_ptr<unsigned char[]> buf_{new unsigned char[kBufSize]};
    std::size_t pos_ = 0;
    std::size_t len_ = 0;
};

}  // namespace

int main() {
    StdinReader in;
    const std::int64_t w = in.readI64();
    const std::int64_t h = in.readI64();
    const std::int64_t maxiter = in.readI64();

    std::vector<std::uint8_t> img(static_cast<std::size_t>(w * h));
    std::int64_t acc = 0;

    for (std::int64_t y = 0; y < h; ++y) {
        for (std::int64_t x = 0; x < w; ++x) {
            const double cr = static_cast<double>(x) / static_cast<double>(w) * 3.5 - 2.5;
            const double ci = static_cast<double>(y) / static_cast<double>(h) * 2.0 - 1.0;
            double zr = 0.0, zi = 0.0;
            std::int64_t it = 0;
            while (it < maxiter) {
                const double t = zr * zr - zi * zi + cr;
                zi = 2.0 * zr * zi + ci;
                zr = t;
                if (zr * zr + zi * zi > 4.0) break;
                ++it;
            }
            img[static_cast<std::size_t>(y * w + x)] = static_cast<std::uint8_t>(it & 255);
            acc += it;
        }
    }

    std::printf("mandel=%lld img0=%d\n", static_cast<long long>(acc),
                static_cast<int>(img[0]));
    return 0;
}

File: bench/bigmatrix/src/mandel.rs
// Shape 4: mandel -- Rust port of mandel.c.
// Read `w h maxiter`; same loop as bench/heavy/mb.c (f64, escape at |z| > 2).

use std::io::{self, Read};

/* ---------- fast stdin: a port of fastio.h ---------- */

/// 1 MiB read buffer plus hand-written integer parsing.
struct FastIn<R: Read> {
    r: R,
    buf: Box<[u8]>,
    len: usize,
    pos: usize,
}

impl<R: Read> FastIn<R> {
    fn new(r: R) -> Self {
        FastIn {
            r,
            buf: vec![0u8; 1 << 20].into_boxed_slice(),
            len: 0,
            pos: 0,
        }
    }

    /// One byte, or -1 at end of input (bm_get).
    #[inline]
    fn get(&mut self) -> i32 {
        if self.pos >= self.len {
            self.len = self.r.read(&mut self.buf).unwrap_or(0);
            self.pos = 0;
            if self.len == 0 {
                return -1;
            }
        }
        let c = self.buf[self.pos] as i32;
        self.pos += 1;
        c
    }

    /// Signed integer, skipping ' ', '\n', '\t', '\r' (bm_i64).
    #[inline]
    fn i64(&mut self) -> i64 {
        let mut c = self.get();
        let mut neg = false;
        while c == b' ' as i32 || c == b'\n' as i32 || c == b'\t' as i32 || c == b'\r' as i32 {
            c = self.get();
        }
        if c == b'-' as i32 {
            neg = true;
            c = self.get();
        }
        let mut v: i64 = 0;
        while (b'0' as i32..=b'9' as i32).contains(&c) {
            v = v.wrapping_mul(10).wrapping_add((c - b'0' as i32) as i64);
            c = self.get();
        }
        if neg {
            -v
        } else {
            v
        }
    }
}

/* ---------- shape ---------- */

fn main() {
    let mut r = FastIn::new(io::stdin().lock());
    let w = r.i64();
    let h = r.i64();
    let maxiter = r.i64();

    let mut img = vec![0u8; (w * h) as usize];
    let mut acc = 0i64;
    for y in 0..h {
        for x in 0..w {
            let cr = x as f64 / w as f64 * 3.5 - 2.5;
            let ci = y as f64 / h as f64 * 2.0 - 1.0;
            let (mut zr, mut zi) = (0.0f64, 0.0f64);
            let mut it = 0i64;
            while it < maxiter {
                let t = zr * zr - zi * zi + cr;
                zi = 2.0 * zr * zi + ci;
                zr = t;
                if zr * zr + zi * zi > 4.0 {
                    break;
                }
                it += 1;
            }
            img[(y * w + x) as usize] = (it & 255) as u8;
            acc += it;
        }
    }
    println!("mandel={} img0={}", acc, img[0]);
}
